[random] support key dtype in custom_vjp

jakevdp opened the pull request:
Fixes #18442

Part of #9263

NeilGirdhar commented:
Awesome, thanks for the quick fix!!

froystig commented on tests/random_test.py:
```diff
@@ -1107,6 +1107,20 @@ class A: pass
+    def _f_fwd(_, state):
+      return state, None
+    def _f_bwd(_, state_bar):
+      return state_bar
```
What does the key part of `state_bar` look like here? A valid key array backed by 0 values?

jakevdp replied:
Yes, it's a valid key array backed by zeros.

froystig replied:
Seems unsafe, no? We don't want such a key to be used.

Tagging @mattjj because we've pondered this in the past.

This may also relate to `jnp.zeros_like` (which implements a stopgap at the moment, arguably), but not necessarily.

jakevdp replied:
If we get key reuse working, it could be a consumed key 😁

jakevdp replied:
OK, changed this to a special key tangent type, so it can't be accidentally used as a key. It will fail very ungracefully if it's used for anything, but we can figure that out later.

froystig commented on tests/random_test.py:
```diff
@@ -1107,6 +1107,20 @@ class A: pass
+      return state, None
+    def _f_bwd(_, state_bar):
+      return state_bar
+    f.defvjp(_f_fwd, _f_bwd)
```
What's the behavior if `symbolic_zeros=True`?

jakevdp replied:
This results in `TypeError: Value CustomVJPPrimal(value=8.0, perturbed=False) with type <class 'jax._src.custom_derivatives.CustomVJPPrimal'> is not a valid JAX type` somewhere in the custom vjp internals.

froystig replied:
Ah, that's because the test returns it as-is from the forward. We'd have to write a slightly different test to exercise symbolic zeros mode, maybe where the forward function tree-maps `lambda x: x.value` over its current output.

The reason to test this mode is that we expect RNG keys to be symbolically zero.

jakevdp replied:
Test is added, thanks!

froystig reviewed (approved):
Nice!

jakevdp committed:
[random] support key dtype in custom_jvp

To do this, we introduce a dtype for key tangents which cannot be used
to generate random values
